kotlinx-coroutines-core/common/src/sync/Mutex.kt

```diff
@@ -117,307 +134,76 @@ public suspend inline fun <T> Mutex.withLock(owner: Any? = null, action: () -> T
 
-    private abstract inner class LockWaiter(
-        @JvmField val owner: Any?
-    ) : LockFreeLinkedListNode(), DisposableHandle {
-        private val isTaken = atomic(false)
-        fun take(): Boolean = isTaken.compareAndSet(false, true)
-        final override fun dispose() { remove() }
-        abstract fun tryResumeLockWaiter(): Boolean
-        abstract fun completeResumeLockWaiter()
-    }
-
-    private inner class LockCont(
-        owner: Any?,
-        private val cont: CancellableContinuation<Unit>
-    ) : LockWaiter(owner) {
-
-        override fun tryResumeLockWaiter(): Boolean {
-            if (!take()) return false
-            return cont.tryResume(Unit, idempotent = null) {
-                // if this continuation gets cancelled during dispatch to the caller, then release the lock
-                unlock(owner)
-            } != null
+    private inner class CancellableContinuationWithOwner(
+        val cont: CancellableContinuation<Unit>,
+        val owner: Any?
+    ) : CancellableContinuation<Unit> by cont {
+        override fun tryResume(value: Unit, idempotent: Any?, onCancellation: ((cause: Throwable) -> Unit)?): Any? {
+            val token = cont.tryResume(value, idempotent, onCancellation)
+            if (token !== null) this@MutexImpl.owner.value = owner
+            return token
         }
 
-        override fun completeResumeLockWaiter() = cont.completeResume(RESUME_TOKEN)
-        override fun toString(): String = "LockCont[$owner, ${cont}] for ${this@MutexImpl}"
-    }
-
-    private inner class LockSelect<R>(
-        owner: Any?,
-        @JvmField val select: SelectInstance<R>,
-        @JvmField val block: suspend (Mutex) -> R
-    ) : LockWaiter(owner) {
-        override fun tryResumeLockWaiter(): Boolean = take() && select.trySelect()
-        override fun completeResumeLockWaiter() {
-            block.startCoroutineCancellable(receiver = this@MutexImpl, completion = select.completion) {
-                // if this continuation gets cancelled during dispatch to the caller, then release the lock
-                unlock(owner)
-            }
-        }
-        override fun toString(): String = "LockSelect[$owner, $select] for ${this@MutexImpl}"
-    }
-
-    // atomic unlock operation that checks that waiters queue is empty
-    private class UnlockOp(
-        @JvmField val queue: LockedQueue
-    ) : AtomicOp<MutexImpl>() {
-        override fun prepare(affected: MutexImpl): Any? =
-            if (queue.isEmpty) null else UNLOCK_FAIL
-
-        override fun complete(affected: MutexImpl, failure: Any?) {
-            val update: Any = if (failure == null) EMPTY_UNLOCKED else queue
-            affected._state.compareAndSet(this, update)
+        override fun resume(value: Unit, onCancellation: ((cause: Throwable) -> Unit)?) {
+            this@MutexImpl.owner.value = owner
+            cont.resume(value, onCancellation)
```

Comment:
What will happen if this resume will have no effect, e.g. because the continuation was concurrently cancelled?